cpp/src/io/parquet/rle_stream.cuh

```diff
@@ -56,92 +55,115 @@ inline __device__ uint32_t get_vlq32(uint8_t const*& cur, uint8_t const* end)
+ * @param level_run RLE run header
+ * @param run_start beginning of data for RLE run
+ * @param end pointer to the end of data for RLE run
+ * @param run_output_pos absolute output position for this run
+ * @param run_offset offset after run_output_pos this call to decode starts outputting at
+ * @param size length that will be decoded in this decode call, truncated to fit output buffer
+ * @param level_bits bits needed to encode max values in the run (definition, dictionary)
+ * @param lane warp lane that is executing this decode call
+ */
+template <typename level_t, int max_output_values>
+__device__ inline void decode(level_t* const output,
+                              int const level_run,
+                              uint8_t const* const run_start,
+                              uint8_t const* const end,
+                              int const run_output_pos,
+                              int const run_offset,
+                              int const size,
+                              int level_bits,
+                              int lane)
+{
+  // local output_pos for this `decode` call.
+  int decode_output_pos = 0;
+  int remain            = size;
 
-  __device__ inline void decode(uint8_t const* const end, int level_bits, int lane, int warp_id)
-  {
-    int output_pos = 0;
-    int remain     = size;
-
-    // for bitpacked/literal runs, total size is always a multiple of 8. so we need to take care if
-    // we are not starting/ending exactly on a run boundary
-    uint8_t const* cur;
-    if (level_run & 1) {
-      int const effective_offset = cudf::util::round_down_safe(run_offset, 8);
-      int const lead_values      = (run_offset - effective_offset);
-      output_pos -= lead_values;
-      remain += lead_values;
-      cur = run_start + ((effective_offset >> 3) * level_bits);
-    }
+  // for bitpacked/literal runs, total size is always a multiple of 8. so we need to take care if
+  // we are not starting/ending exactly on a run boundary
+  uint8_t const* cur;
+  if (is_literal_run(level_run)) {
+    int const effective_offset = cudf::util::round_down_safe(run_offset, 8);
+    int const lead_values      = (run_offset - effective_offset);
+    decode_output_pos -= lead_values;
+    remain += lead_values;
+    cur = run_start + ((effective_offset >> 3) * level_bits);
+  }
 
-    // if this is a repeated run, compute the repeated value
-    int level_val;
-    if (!(level_run & 1)) {
-      level_val = run_start[0];
-      if (level_bits > 8) { level_val |= run_start[1] << 8; }
+  // if this is a repeated run, compute the repeated value
+  int level_val;
+  if (is_repeated_run(level_run)) {
+    level_val = run_start[0];
+    if constexpr (sizeof(level_t) > 1) {
+      if (level_bits > 8) {
+        level_val |= run_start[1] << 8;
+        if constexpr (sizeof(level_t) > 2) {
+          if (level_bits > 16) {
+            level_val |= run_start[2] << 16;
+            if constexpr (sizeof(level_t) > 3) {
```

Comment:
How are these two `constexpr` conditions different?

Reply:
Good point.

Reply:
Removed the extraneous check.